File: server/haskell/src/Record/Undo.hs
module Record.Undo (undo) where

import Auth.JWT (authorizedUserId)
import DB (execute)
import Data.Aeson.TH (defaultOptions, deriveJSON)
import Data.Int (Int64)
import qualified Data.Set as Set
import qualified Data.Text as T
import qualified Data.Vector as V
import qualified Hasql.Pool as HP
import qualified Hasql.Statement as S
import qualified Hasql.TH as TH
import Network.HTTP.Types (status200, status403, status500)
import Record.Go (Color (Black, White), playStones, runBoardA, toCoord')
import Web.Scotty
  ( ScottyM,
    json,
    param,
    post,
    raiseStatus,
    status,
  )

data Stone = Stone {color :: T.Text, x :: Int, y :: Int} deriving (Eq, Show)

$(deriveJSON defaultOptions ''Stone)

data Point = Point {x :: Int, y :: Int} deriving (Eq, Show)

$(deriveJSON defaultOptions ''Point)

data UndoResponse = UndoResponse {add :: [Stone], remove :: [Point]} deriving (Eq, Show)

$(deriveJSON defaultOptions ''UndoResponse)

selectRecord :: S.Statement (Int64, Int64) Int64
selectRecord =
  [TH.singletonStatement|
    select
    board_size :: int8
    from record_record
    where
    owner_id = $1 :: int8 and id = $2 :: int8
  |]

type MoveRow = (Int64, Maybe Int64, T.Text)

selectMoves :: S.Statement Int64 (V.Vector MoveRow)
selectMoves =
  [TH.vectorStatement|
    select
    move :: int8, position :: int8?, color :: text
    from record_move
    where
    record_id = $1 :: int8
    order by move asc
  |]

deleteMove :: S.Statement (Int64, Int64) Int64
deleteMove =
  [TH.rowsAffectedStatement|
    delete from record_move where
    record_id = $1 :: int8 and move = $2 :: int8
    |]

toColor :: T.Text -> Color
toColor color' = case color' of
  "B" -> Black
  "W" -> White
  _ -> White -- TODO????

fromColor :: Color -> T.Text
fromColor Black = "B"
fromColor White = "W"

invert :: Color -> Color
invert Black = White
invert White = Black

undo :: HP.Pool -> ScottyM ()
undo pool = post "/api/records/:recordId/undo/" $ do
  userId <- authorizedUserId
  recordId <- param "recordId"
  size' <- execute pool selectRecord (userId, recordId)
  moves' <- execute pool selectMoves recordId
  let size = fromIntegral size'
      (moveNumber, pos, color) = V.last moves'
      removals = case pos of
        Just pos' -> do
          let (x, y) = toCoord' size $ fromIntegral pos'
          [Point {x, y}]
        Nothing -> []
  case [(fmap fromIntegral pos', toColor color') | (_, pos', color') <- V.toList moves'] of
    [] -> raiseStatus status403 "No moves to undo"
    moves -> do
      case runBoardA size $ playStones moves of
        Right (_, captures) -> do
          let additions =
                [ do
                    let (x, y) = toCoord' size pos'
                    Stone {x, y, color = fromColor $ invert $ toColor color}
                  | pos' <- Set.toAscList captures
                ]
          _ <- execute pool deleteMove (recordId, moveNumber)
          status status200
          json UndoResponse {add = additions, remove = removals}
        Left _ -> raiseStatus status500 "Error replaying game"

File: server/rust/src/record/undo.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio_postgres::Client;

use crate::auth::UserId;
use crate::db::{query, query_one};
use crate::derow;
use crate::record::go::{find_captures, Color, GoError, Move, Pos};

#[derive(Serialize, Deserialize)]
struct Stone {
    color: String,
    x: i32,
    y: i32,
}

#[derive(Serialize, Deserialize)]
struct Point {
    x: i32,
    y: i32,
}

#[derive(Serialize, Deserialize)]
struct UndoResponse {
    add: Vec<Stone>,
    remove: Vec<Point>,
}

pub async fn undo(
    State(client): State<Arc<Client>>,
    UserId(user_id): UserId,
    Path(record_id): Path<i64>,
) -> impl IntoResponse {
    let record_row = query_one(
        &client,
        "SELECT board_size FROM record_record WHERE owner_id=$1 AND id=$2",
        &[&user_id, &record_id],
    )
    .await?;
    let board_size = record_row.get("board_size");
    let mut moves: Vec<Move> = query(
        &client,
        "SELECT position, color FROM record_move WHERE record_id=$1 ORDER BY move ASC",
        &[&record_id],
    )
    .await?
    .iter()
    .map(|row| derow!(row, {position: Option<i32>, color: &str}))
    .map(|(pos, color)| (pos.map(Pos::new), Color::new(color)))
    .collect();
    if moves.is_empty() {
        return Err((StatusCode::FORBIDDEN, "No moves to undo"));
    }
    let move_to_undo = moves.pop().unwrap();
    // The most recent move needs to be removed from the board
    let removals = if let (Some(pos), _) = &move_to_undo {
        let (x, y) = pos.to_coord(board_size);
        vec![Point { x, y }]
    } else {
        vec![]
    };
    // Any stones that move captured need to be added back to the board
    // They will have the opposite color from the move being undone
    let addition_color = format!("{}", move_to_undo.1.invert().clone());
    let additions = match find_captures(board_size, &moves, &move_to_undo) {
        Err(GoError::OutOfBounds(_)) => {
            return Err((StatusCode::FORBIDDEN, "Out of bounds"));
        }
        Err(GoError::SpaceOccupied(_)) => {
            return Err((StatusCode::FORBIDDEN, "Already a stone there"));
        }
        Err(GoError::Suicide(_)) => {
            return Err((StatusCode::FORBIDDEN, "Move is suicidal"));
        }
        Ok(captures) => captures
            .iter()
            .map(|pos| pos.to_coord(board_size))
            .map(|(x, y)| Stone {
                x,
                y,
                color: addition_color.clone(),
            })
            .collect(),
    };
    let move_number: i32 = (1 + moves.len()).try_into().unwrap();
    query(
        &client,
        "DELETE FROM record_move WHERE record_id=$1 AND move=$2",
        &[&record_id, &move_number],
    )
    .await?;
    Ok((
        StatusCode::OK,
        Json(UndoResponse {
            add: additions,
            remove: removals,
        }),
    ))
}
